backends/tfhe-cuda-backend/cuda/src/integer/integer.cuh

```diff
@@ -627,26 +627,46 @@ void rotate_left(Torus *buffer, int mid, uint32_t array_length) {
   std::rotate(buffer, buffer + mid, buffer + array_length);
 }
 
+/// Caller needs to ensure that the operation applied is coherent from an
+/// encoding perspective.
+///
+/// For example:
+///
+/// Input encoding has 2 bits and output encoding has 4 bits, applying the
+/// identity lut would map the following:
+///
+/// 0|00|xx -> 0|00|00
+/// 0|01|xx -> 0|00|01
+/// 0|10|xx -> 0|00|10
+/// 0|11|xx -> 0|00|11
+///
+/// The reason is the identity function is computed in the input space but the
+/// scaling is done in the output space, as there are more bits in the output
+/// space, the delta is smaller hence the apparent "division" happening.
 template <typename Torus>
-void generate_lookup_table(Torus *acc, uint32_t glwe_dimension,
-                           uint32_t polynomial_size, uint32_t message_modulus,
-                           uint32_t carry_modulus,
-                           std::function<Torus(Torus)> f) {
-
-  uint32_t modulus_sup = message_modulus * carry_modulus;
-  uint32_t box_size = polynomial_size / modulus_sup;
-  Torus delta = (1ul << 63) / modulus_sup;
+void generate_lookup_table_with_encoding(Torus *acc, uint32_t glwe_dimension,
+                                         uint32_t polynomial_size,
+                                         uint32_t input_message_modulus,
+                                         uint32_t input_carry_modulus,
+                                         uint32_t output_message_modulus,
+                                         uint32_t output_carry_modulus,
+                                         std::function<Torus(Torus)> f) {
+
+  uint32_t input_modulus_sup = input_message_modulus * input_carry_modulus;
+  uint32_t output_modulus_sup = output_message_modulus * output_carry_modulus;
+  uint32_t box_size = polynomial_size / input_modulus_sup;
+  Torus output_delta = std::numeric_limits<Torus>::max()  / (2 * output_modulus_sup);
```

Comment:
@IceTDrinker this is not exactly `(1ul << 63) / modulus_sup` but it is close. `std::numeric_limits<Torus>::max() / 2` should be smaller than `1<<63` by one. Do you think that's ok?

Reply:
another way to do it is 1 << (Torus::BITS - 1)

Reply:
`static_cast<Torus>(1) << (Torus::BITS - 1)` or `Torus(1) << (Torus::BITS - 1)`

as I think in `1 << (Torus::BITS - 1)` the left most constant `1` is going to be an int as all C constant that do not have postfix are int

Reply:
yes of course, just meant the idea was the above, but you are right to precise it

Reply:
to be fully precise it would be

```(static_cast<Torus>(1) << (Torus::BITS - 1)) / output_modulus_sup``` in this exact line you commented

Reply:
Just to check you have Torus::BITS defined in CUDA ? I was using pseudo code